guava/src/com/google/common/escape/UnicodeEscaper.java

```diff
@@ -127,7 +130,7 @@ public String escape(String string) {
    * @throws IllegalArgumentException if the scanned sub-sequence of {@code csq} contains invalid
    *     surrogate pairs
    */
-  protected int nextEscapeIndex(CharSequence csq, int start, int end) {
+  protected @NonNegative int nextEscapeIndex(CharSequence csq, @IndexFor("#1") int start, @IndexFor("#1") int end) {
```

Comment:
Seems like there are some suppressed warnings related to this method, so let's look at it.

First, the documentation says that `end` is an index immediately after a character (in `csq`). Right from that, we can expect that `end` can be equal to `csq.length()` (an index after the last character). Therefore, `@IndexFor("#1")` is too strong for `end`, because it doesn't allow it to be `csq.length()`. To allow that, use `@IndexOrHigh("#1")`.

Second, the documentation states that `start` is an index of a character. That seems to suggest, that `@IndexFor` is the right annotation for it. However, when a method takes a sequence and two indices that specify a subsequence within that sequence, usually empty sequences are also allowed.
(As an example of that, consider `String.substring`. These calls `"abc".substring(0, 0)`, `"abc".substring(0, 3)`, `"abc".substring(3, 3)`, and `"".substring(0, 0)` are all valid.)
Is it the same case here? When the documentation does not provide a clear answer (often the case with non-public methods), it's helpful to look at the implementation and usage of the method, to see what is the intended behavior.
- The method is called from `UnicodeEscaper.escape(String)`.  Users can call `escape("")`, which will in turn call `nextEscapeIndex("", 0, 0)`. From that, we can conclude that `nextEscapeIndex` is supposed to work correctly even if `start == csq.length()`. Therefore, I would argue that `start` should also be annotated `@IndexOrHigh("#1")`.
- Let's look what happens in the implementation. When both indices are the same, the while loop is skipped and the method simply returns `start`. So it will behave correctly.

Last, the method returns an index. It would be good to express that with an annotation. Clearly, the returned value can be equal to the length of the sequence. So again, `@IndexOrHigh("#1")` is the right annotation for the return type.

Reply:
@NhatDinh, do you agree with this reasoning? If yes, can you change the annotations on this method and other methods where it applies, so that we can proceed with the review?

Reply:
@panacekcz I agree with your reasoning, but like method codePointAt(), this line `index += Character.isSupplementaryCodePoint(cp) ? 2 : 1;` will cause a `compound.assignment.type.incompatible` warning. I think we need to discuss it more in our meeting.